blog/06-02-2023-catering-to-the-client-side.md

```diff
@@ -0,0 +1,173 @@
+
+```mermaid
+sequenceDiagram
+  actor app
+  participant client as feature flagging logic
+  participant cache as in-memory cache
+  participant provider as flag service
+  app->>+client: userLoggedIn(userId)
+  client->>+provider: evaluateFlags(evaluationContext)
+  provider-->>-client: flag values
+  client->>+cache: store(flag values)
+  cache->>-client: .
+  %% deactivate cache
+  client->>-app: .
+  %% deactivate client
+  note right of app: some time later..
+  app->>+client: operation that needs a flagging decision
+  client->>+cache: getFlagValue(flagKey)
+  note right of cache: no call to flagging service needed
+  cache->>-client: previously evaluated value
+  client->>-app: .
+
+
+```
+
+Put another way, we seperate flag [**evaluation**](/docs/specification/glossary#evaluating-flag-values) - passing an evaluation context through a set of rules in order to determine a flagging decision - from flag [**resolution**](/docs/specification/glossary#resolving-flag-values) - getting the flagging decision for a specific feature flag.
+
+## Client-side support in OpenFeature
+
+With OpenFeature we have been thinking about how to support these key differences between client-side and server-side feature flagging. We think about these differences as two paradigms: dynamic context (for server-side flags) and static context (for client-side flags). 
+
+OpenFeature's current Evaluation API supports the dynamic paradigm quite nicely, but to support the static paradigm (and thus client-side flagging) we need to add a second flavor of the Evaluation API. 
+
+### Server-side evaluation today
+Let's compare and contrast. A typical server-side flagging decision using OpenFeature's current SDK might look something like this:
+``` java
+@GetMapping("/hello")
+public String getSalutation() {
+    final Client client = openFeatureAPI.getClient();
+    final evalContext:EvaluationContext = evalContextForCurentRequest();
+
+    if (client.getBooleanValue("use-formal-salutation", false, evalContext)) {
+      return "Good day to you!";
+    }else{
+      return "Hey, what's up?";
+    }
+} 
+```
+You can see that we're passing evaluation context every time we need to make a flagging decision.
+
+### Client-side evaluation tomorrow
+With the currently proposed OpenFeature changes, a client-side flagging decision would look more like this:
+
+``` java
+public string generateSalutation(){
+  if (client.getBooleanValue("use-formal-salutation", false)) {
+    return "Good day to you!";
+  }else{
+    return "Hey, what's up?";
+  }
+}
+```
+
+You can see we are no longer passing any evaluation context when requesting a flagging decisions. 
+
+But our flagging framework needs to get flag evaluation context from somewhere. So rather than passing the context as part of each flagging decision, the app provides a new evaluation context to OpenFeature whenever a user logs in or out (as part of this app's hypothetical `onAuthenticated(...)` handler): 
```

Comment:
This makes it sound like the only place that context can change is during log in or out. Could you try and make it clear that this is just an example of one scenario where context may be updated?

Reply:
I gave this a try, but couldn't figure out a way to make this clear without distracting the reader from the main point - how an app might set evaluation context as a side effect of some context change within the app

Reply:
I gave this a try, but couldn't figure out a way to make this clear without distracting the reader from the main point - how an app might set evaluation context as a side effect of some context change within the app